### ISSUE TITLE
Examples broken on AMD display device

### ISSUE THREAD

**Issue by pavanky** [arrayfire/forge#94]
OpenCL examples are not working at all

**surface**
![surface_opencl](https://cloud.githubusercontent.com/assets/364829/16026082/bab75562-3199-11e6-8fa5-3e4a02818494.png)

**plotting**
![plotting_opencl](https://cloud.githubusercontent.com/assets/364829/16026080/bab52f9e-3199-11e6-8ed8-8f78b64f2dd3.png)

**plot3**
![plot3_opencl](https://cloud.githubusercontent.com/assets/364829/16026084/bab84c42-3199-11e6-8316-de4a8c01184e.png)

**histogram**
![histogram_opencl](https://cloud.githubusercontent.com/assets/364829/16026083/bab7655c-3199-11e6-89cf-75651ac3ede7.png)

**fractal**
![fractal_opencl](https://cloud.githubusercontent.com/assets/364829/16026081/bab6fbda-3199-11e6-80d9-f8f9620b2e6e.png)

---------------------------

CPU examples look garbled at default window size, but they work after maximizing the window.

**Histogram default window**
![histogram_cpu_default](https://cloud.githubusercontent.com/assets/364829/16026108/e7ceee66-3199-11e6-93ba-db48a2c27e6f.png)

**Histogram maximized window**
![histogram_cpu_maximized](https://cloud.githubusercontent.com/assets/364829/16026107/e7cdfdb2-3199-11e6-820a-bcb09dfb8e87.png)

**Fractal default window**
![fractal_cpu_default](https://cloud.githubusercontent.com/assets/364829/16026110/e7d2a038-3199-11e6-95dc-71dee8507228.png)

**Fractal maximized window**
![fractal_cpu_maximized](https://cloud.githubusercontent.com/assets/364829/16026109/e7d136e4-3199-11e6-8879-970f3564c5a6.png)

**Comment by pavanky** [comment]
OK you can't see the image being garbled in the screen shot of fractal default size, but it is garbled. The same is true for all cpu graphics examples on my AMD GPU

**Comment by 9prady9** [comment]
@pavanky I have checked the examples on the #92 PR changes and everything is rendering as expected on AMD APU A10. Can you please test this on your GPU and let me know if the problem is still there after the changes of PR #92

**Comment by 9prady9** [comment]
@pavanky I have been able to finally setup linux environment (the details are given below) and opencl examples from devel branch run with few glitches but are not as corrupted as we see above.

System Info
```bash
pradeep@asgard ~/gitroot/forge (devel) $uname -a
Linux asgard 4.4.0-21-generic #37-Ubuntu SMP Mon Apr 18 18:33:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
Ubuntu 16.04
```
GPU Info
```bash
  Platform ID:					 0x7fa1444e98f8
  Name:						 Spectre
  Vendor:					 Advanced Micro Devices, Inc.
  Device OpenCL C version:			 OpenCL C 1.2 
  Driver version:				 2117.7 (VM)
  Profile:					 FULL_PROFILE
  Version:					 OpenCL 1.2 AMD-APP (2117.7)
  Extensions:					 cl_khr_fp64 cl_amd_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_gl_sharing cl_amd_device_attribute_query cl_amd_vec3 cl_amd_printf cl_amd_media_ops cl_amd_media_ops2 cl_amd_popcnt cl_khr_image2d_from_buffer cl_khr_spir cl_khr_gl_event 
```
GPU Driver Info
```
AMDGPU-Pro Driver Version 16.30 for Ubuntu 16.04
```

Glitches i see on this machine are as follows

* Renderings appear corrupted `while` resizing the window although they render fine before and after resize operation.
* Towards end of the program, when we close the window, the opencl examples report a `seg fault` even though it ran fine until that point.

**Comment by pavanky** [comment]
Still broken on devel. The problem occurs even if I do not resize the window. It disappears after I perform any resizing operation.

**Comment by pavanky** [comment]
@9prady9 The examples are not broken on AMD when using SDL2 (as opposed to GLFW3). So I am guessing the problem occurs somewhere in the window focus / resizing in forge when using glfw.

But the responsiveness is really really bad on SDL. I am going to check if the responsiveness is bad on other GPUs.

**Comment by 9prady9** [comment]
@pavanky I tried SDL2 toolkit backend on my laptop that has GTX 650M and image examples(fractal, histogram) & surface example were responding smoothly for the mouse events without lag. Could it be possible that this is another issue on AMD hardware ? do you see the lag for SDL2 and cpu examples as well ?

**Comment by pavanky** [comment]
@9prady9 no, the problem occurs on other gpus as well: https://github.com/arrayfire/forge/issues/109

**Comment by pavanky** [comment]
I am trying this on arch, I am assuming you are doing the same @9prady9 ?

**Comment by 9prady9** [comment]
no, i am using ubuntu 16.04 now

**Comment by 9prady9** [comment]
![IMG_20210504_224922](https://user-images.githubusercontent.com/3270458/117045455-2886f980-ad2d-11eb-9ea5-b33302b40999.jpg) - AMD GPU rendering seems to work in general except the 3d renderables  - an odd horizontal line appears from the center of the screen to left edge of the window and it keeps rotation with the object if the object rotates - doesn't matter what the rendered object is. Static image of that horizontal line is shown below
![IMG_20210504_225502](https://user-images.githubusercontent.com/3270458/117045463-2c1a8080-ad2d-11eb-87f2-68bd70e56417.jpg)
